## **COURSE NAME AND CODE:** Programming for Science Informatics - B573

### **Programmer:** Sharmila Tummala

Date: 09/26/2024

### **Programming Language:** 
Python

### **Python version:** 
Python 3.10.12

### **Description:** 
I am submitting a python script that allows to read DNA sequence from chr1_GL383518v1_alt file and retrieve the letters in the asked position of the sequence, get its Reverse complement sequence and retrieve letters as per the requirement, counting each base occurance in every 1000 nucleotide subsequence of the DNA and storing it in a nested Dictionary, access the dictionary and creating a list of count of bases in each sub sequence and adding them all. This python script helps to understand nested dictionaries and nested lists and how to access elements in them. I have changes all the letters into uppercase in the program to make it easy to count the bases.


### Files needed:
you need python installed and "chr1_GL383518v1_alt" DNA sequence file from UCSC Genome Browser https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/


### Packages needed: 
This is a basic python program it works without any special packages.

### How to run my program
1. download the a2.py file and chr1_GL383518v1_alt to your local system
2. open the terminal and navigate to the directory which contains these files using \'cd path\' command
3. enter \"python a2.py\" to run script

if you are using VScode
1. make sure the python extension is installed
2. download this file and open it using VScode and run the program by clicking on the run button present on the top right corner

### NOTE
1. Python has various IDEs, running this python file completely depends on the IDE you are using.
2. Keep python file and the DNA seqence file in the same folder.
3. If you can't access the file please give file path where the downloaded sequence file is located in your system in the place of file name in the first line of the code like this.  
**with open('/content/drive/MyDrive/Colab Notebooks/chr1_GL383518v1_alt.fa', 'r')as file:**




